### PR TITLE
Add `XML::Reader#finalize`

### PR DESCRIPTION
**Pull request by straight-shoota**
Adds a finalize handler that releases libXML resources

### CODE REVIEW

File: src/xml/reader.cr [hunk 1]
```diff
@@ -28,6 +28,13 @@ class XML::Reader
     )
   end
 
+  def finalize
+    if reader = @reader
+      LibXML.xmlFreeTextReader(reader)
+      @reader = Pointer(Void).null.as(LibXML::XMLTextReader)
```
Comment (asterite) [inline src/xml/reader.cr:34]: Is this last line needed? 🤔

Comment (straight-shoota) [reply src/xml/reader.cr:34]: Trying to avoid errors if `finalize` is called multiple times.

Comment (straight-shoota) [reply src/xml/reader.cr:34]: Not sure if it's strictly necessary, though. If not, we don't need the condition either.

Comment (asterite) [reply src/xml/reader.cr:34]: I don't think `finalize` could be called manually.

Comment (asterite) [reply src/xml/reader.cr:34]: could -> should

Comment (straight-shoota) [reply src/xml/reader.cr:34]: But it can 🤷

Comment (asterite) [reply src/xml/reader.cr:34]: But it shouldn't!

Comment (asterite) [reply src/xml/reader.cr:34]: If you do that, well, face the consequences.

Comment (lbguilherme) [reply src/xml/reader.cr:34]: Please, either make it impossible to misuse (compiler error if called) or safe to misuse (noop).

Comment (asterite) [reply src/xml/reader.cr:34]: Okay. Then prevent calling `finalize` and continuing using the object too while at it ;-)

Comment (straight-shoota) [reply src/xml/reader.cr:34]: Can we merge this as is? It's better to be extra safe than sorry. The discussion about being able to call `finalize` is a different one (I guess #6331).

Comment (HertzDevil) [reply src/xml/reader.cr:34]: There is no concensus on whether `#finalize` could anticipate multiple calls, even within the same module:

https://github.com/crystal-lang/crystal/blob/1c33a0b1d14a1e2c8835bb1b7381789d648624ea/src/llvm/pass_manager_builder.cr#L38-L40

https://github.com/crystal-lang/crystal/blob/1c33a0b1d14a1e2c8835bb1b7381789d648624ea/src/llvm/pass_builder_options.cr#L22-L27

The latter was added in #12116 ;-)

**Comment by HertzDevil** [comment]
Is `XML::Reader` the only type that needs extra finalization? Why don't we need to call other deallocation functions like `xmlFreeNode`?

**Comment by straight-shoota** [comment]
@HertzDevil I've only looked at parsing and there it's the only one that needs to free resources.

But I've recently noticed that libxml2 is hooked up with the garbage collector. So probably we don't need `finalize`.

**Comment by HertzDevil** [comment]
I don't fully understand how libxml2 is connected to the garbage collector. But even if that were the case, a leaked `XML::Node` for example would only call `GC.free`, and that might be insufficient to clean up any resources used by the node. That's why I wonder if we need `xmlFreeNode`.

The docs say that `xmlFreeNode` deletes subnodes recursively, maybe that's why we don't need it?

**Review by beta-ziliani — APPROVED** [review]

**Comment by straight-shoota** [comment]
I'm not sure anymore if this actually makes sense.

**Pull request closed without merging**